File: AspNetMvcUnleashed/0672329980_source/Chapter11/VB/MvcApplication1/MvcApplication1/Models/SimpleMovieService.vb
Imports MvcFakes

Public Class SimpleMovieService
    Implements ISimpleMovieService

    Private _repository As ISimpleMovieRepository
    Private _cache As ICache

    Public Sub New()
        Me.New(New SimpleMovieRepository(), New CacheWrapper(HttpContext.Current.Cache))
    End Sub

    Public Sub New(ByVal repository As ISimpleMovieRepository, ByVal cache As ICache)
        _repository = repository
        _cache = cache
    End Sub


    Public Function ListMoviesCached() As IEnumerable(Of Movie) Implements ISimpleMovieService.ListMoviesCached
        Dim movies = CType(_cache("movies"), IEnumerable(Of Movie))
        If movies Is Nothing Then
            movies = ListMovies()
            _cache("movies") = movies
        End If
        Return movies
    End Function


    Public Function ListMovies() As IEnumerable(Of Movie)
        Return _repository.ListMovies()
    End Function

End Class

File: AspNetMvcUnleashed/0672329980_source/Chapter11/CS/MvcApplication1/MvcApplication1/Models/SimpleMovieService.cs
using System.Collections.Generic;
using System.Web;
using MvcFakes;

namespace MvcApplication1.Models
{
    public class SimpleMovieService : ISimpleMovieService
    {
        private ISimpleMovieRepository _repository;
        private ICache _cache;

        public SimpleMovieService()
            : this(new SimpleMovieRepository(), new CacheWrapper(HttpContext.Current.Cache)) {}

        public SimpleMovieService(ISimpleMovieRepository repository, ICache cache)
        {
            _repository = repository;
            _cache = cache;
        }


        public IEnumerable<Movie> ListMoviesCached()
        {
            var movies = (IEnumerable<Movie>)_cache["movies"];
            if (movies == null)
            {
                movies = ListMovies();
                _cache["movies"] = movies;
            }
            return movies;
        }


        public IEnumerable<Movie> ListMovies()
        {
            return _repository.ListMovies();
        }

    }
}
